File: razer-sdk/src/test/java/com/bonepl/razersdk/effects/LayeredCustomKeyboardEffectTest.java
package com.bonepl.razersdk.effects;

import com.bonepl.razersdk.ChromaRestSDK;
import com.bonepl.razersdk.animation.AnimatedFrame;
import com.bonepl.razersdk.color.Color;
import com.bonepl.razersdk.animation.LayeredFrame;
import com.bonepl.razersdk.animation.SimpleFrame;
import com.bonepl.razersdk.color.StaticColor;
import com.bonepl.razersdk.sdk.RzKey;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

class LayeredCustomKeyboardEffectTest {

    @Test
    @SuppressWarnings("JUnitTestMethodWithNoAssertions")
    void testLayeredKeyboardEffect() throws InterruptedException {
        try (ChromaRestSDK razerSDKClient = new ChromaRestSDK()) {
            final AnimatedFrame firstAnimatedFrame = createFramesFromEnum(7, 18, StaticColor.GREEN);
            final AnimatedFrame secondAnimatedFrame = createFramesFromEnum(0, 11, StaticColor.BLUE);
            for (int i = 0; i <= 100; i += 5) {
                LayeredFrame layeredFrame = new LayeredFrame();
                layeredFrame.addFrame(new SimpleFrame(new StaticColor(30, 30, 0)));
                layeredFrame.addFrame(firstAnimatedFrame);
                layeredFrame.addFrame(secondAnimatedFrame);
                if (i % 10 == 0) {
                    layeredFrame.addFrame(new SimpleFrame(RzKey.RZKEY_SPACE, StaticColor.RED));
                }
                razerSDKClient.createKeyboardEffect(layeredFrame);
                Thread.sleep(100);
            }
        }
    }

    private static AnimatedFrame createFramesFromEnum(int from, int to, Color color) {
        final List<RzKey> keys = Arrays.stream(RzKey.values())
                .skip(from).limit(to)
                .collect(Collectors.toList());
        final AnimatedFrame animatedFrame = new AnimatedFrame();
        for (int i = 0; i < to; i++) {
            animatedFrame.addAnimationFrame(2, new SimpleFrame(keys.subList(0, i), color));
        }
        return animatedFrame;
    }
}

File: test/razer-sdk/effects/LayeredCustomKeyboardEffect.test.ts
import {describe, test} from 'vitest';
import {ChromaNativeSDK} from '../../../src/razer-sdk/ChromaNativeSDK.js';
import {AnimatedFrame} from '../../../src/razer-sdk/animation/AnimatedFrame.js';
import {LayeredFrame} from '../../../src/razer-sdk/animation/LayeredFrame.js';
import {SimpleFrame} from '../../../src/razer-sdk/animation/SimpleFrame.js';
import {Color} from '../../../src/razer-sdk/color/Color.js';
import {StaticColor} from '../../../src/razer-sdk/color/StaticColor.js';
import {getRzKeyValues, RzKey} from '../../../src/razer-sdk/sdk/RzKey.js';

function createFramesFromEnum(from: number, to: number, color: Color): AnimatedFrame {
    const allKeys = getRzKeyValues();
    const keys = allKeys.slice(from, from + to);
    const animatedFrame = new AnimatedFrame();
    for (let i = 0; i < to; i++) {
        animatedFrame.addAnimationFrame(2, new SimpleFrame(keys.slice(0, i), color));
    }
    return animatedFrame;
}

/**
 * Integration test for layered frame composition on actual hardware.
 * Demonstrates multiple animation layers with background and overlay effects.
 */
describe('LayeredCustomKeyboardEffect Integration', () => {
    test('testLayeredKeyboardEffect', async () => {
        const sdk = new ChromaNativeSDK();
        await sdk.connect();

        try {
            const firstAnimatedFrame = createFramesFromEnum(7, 18, StaticColor.GREEN);
            const secondAnimatedFrame = createFramesFromEnum(0, 11, StaticColor.BLUE);

            for (let i = 0; i <= 100; i += 5) {
                const layeredFrame = new LayeredFrame();
                layeredFrame.addFrame(new SimpleFrame(new StaticColor(30, 30, 0)));
                layeredFrame.addFrame(firstAnimatedFrame);
                layeredFrame.addFrame(secondAnimatedFrame);
                if (i % 10 === 0) {
                    layeredFrame.addFrame(new SimpleFrame(RzKey.RZKEY_SPACE, StaticColor.RED));
                }

                await sdk.createKeyboardEffect(layeredFrame);
                await new Promise(resolve => setTimeout(resolve, 100));
            }
        } finally {
            await sdk.close();
        }
    }, 30000);
});
